draft-ietf-taps-interface.md

```diff
@@ -1780,38 +1772,170 @@ that the other endpoint is done sending on a connection.
+- Delimeter-separated protocols, such as HTTP/1.1.
 
-Concretely, receiver-side de-framing allows a caller to provide the interface
-with a function that takes an octet stream, as provided by the underlying
-Protocol Stack, reads and returns a single Message of an appropriate type for
-the application and platform, and leaves the octet stream at the start of the
-next Message to deframe. It consists of a Deframer Object with a single Action,
-Deframe. Since the Deframer depends on the protocol used at the application
-layer, it is bound to the Preconnection during the pre-establishment phase:
+Note that while Message Framers add the most value when placed above
+a protocol that otherwise does not preserve message boundaries, they can
+also be used with datagram- or message-based protocols. In these cases,
+they add an additional transformation to further encode or encapsulate,
+and can potentially support packing multiple application-layer Messages
+into individual transport datagrams.
+
+## Defining Message Framers
+
+Applications can define a Message Framer by creating a object to represent a unique
+definition of a framing protocol. Event handlers for inbound and outbound data
+are registered on this object.
 
 ~~~
-Preconnection.DeframeWith(Deframer)
+framer := NewMessageFramer()
+framer.setEventHandlers()
+~~~
+
+## Adding Message Framers to Connections
 
-{messageData} := Deframer.Deframe(OctetStream)
+The Message Framer object can then be added to one or more Preconnections
+to run on top of transport protocols. Multiple Framers may be added.
+
+~~~
+Preconnection.AddFramer(framer)
+~~~
+
+When sending Messages, applications can add specific Message
+values to a MessageContext ({{message-props}}) that is intended for a Framer.
+This can be used, for example, to set the type of a Message for a TLV protocol.
+The namespace of values is custom for each unique Message Framer.
+
+~~~
+messageContext := NewMessageContext()
+messageContext.add(framer, key, value)
+Connection.Send(messageData, messageContext)
+~~~
+
+When an application receives a MessageContext in a Receive event,
+it can also look to see if a value was set by a specific Message Framer.
+
+~~~
+messageContext.get(framer, key) -> value
+~~~
+
+## Message Framer Lifetime
+
+When a Connection establishment attempt begins, an event is delivered to
+notify the framer that a new Connection is being created and has a new
+instance of a Message Framer. Similarly, a stop event is delivered when a
+Connection is being torn down.
+
+~~~
+FramerInstance -> Start()
+FramerInstance -> Stop()
 ~~~
 
+When Message Framer gets a start message, it sets up its state and then indicates
+to the connection that it is ready to handle sending Messages by calling `ready` on
+the FramerInstance. This call can be delayed in case the protocol needs to do some
+simple handshake exchange before handling application Messages.
+
+~~~
+instance.ready()
```

Comment:
Here, and below, I wonder about this use of "instance". Above you talk about a a FramerInstance and call Start and Stop on it. Why has this FramerInstance only become "instance" all of a sudden?

Reply:
Not sure on our pseudocode parlance, but this was representing the call that an app would make on a local variable that holds their reference to a FramerInstance. Having `FramerInstance.ready()` looks like a "class method" rather than an "instance method".